### PR TITLE
[AEA-1315] p2p libp2p node restart on read/write fail

### LINKED ISSUE #2158: Test case (end-to-end) for checking recovery from network disconnection
**Is your feature request related to a problem? Please describe.**
We should have a test case to check that an agent can autonomously recover from a network disconnection when using `p2p_libp2p` and other `fetchai` authored connections.

**Describe the solution you'd like**
TBD

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

p2p libp2p node restart on read/write fail

## Fixes

#2158

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/fetchai/connections/p2p_libp2p/connection.py [hunk 1]
```diff
@@ -351,8 +364,15 @@ async def read(self) -> Optional[bytes]:
...
-            return None
+            self.logger.exception(
+                f"Failed to read. Exception: {e}. Try reconnect to node and read again."
+            )
+            await self.restart()
+            try:
+                return await self.pipe.read()
+            except Exception:  # pragma: nocover pylint: disable=broad-except
+                self.logger.exception(f"Failed to read. Exception: {e}.")
+                return None
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/connection.py:375]:
```suggestion
                self.logger.exception(f"Failed to read after node restart. Exception: {e}.")
                return None
```

File: packages/fetchai/connections/p2p_libp2p/connection.py [hunk 2]
```diff
@@ -338,7 +343,15 @@ async def write(self, data: bytes) -> None:
...
-        await self.pipe.write(data)
+
+        try:
+            await self.pipe.write(data)
+        except Exception:  # pylint: disable=broad-except
+            self.logger.exception(
+                "Exception raised on message write. Try reconnect to node and write again."
+            )
+            await self.restart()
+            await self.pipe.write(data)
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/connection.py:354]: Should we not catch the second exception here too (like in read)?

Comment (solarw) [reply packages/fetchai/connections/p2p_libp2p/connection.py:354]: no, just follow the original design, and this one does not return a value at all

**Review by marcofavorito — APPROVED** [review]
LGTM

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**